File: Data_structure/836/disjoint_set.cpp
#include <iostream>
using namespace std;

const int N = 1e5 + 10;
int n, m, a, b, pa, pb, p[N];
char op[2];

int find(int x);

int main() {
    scanf("%d%d", &n, &m);
    for (int i = 1; i <= n; i++) p[i] = i;
    while (m--) {
        scanf("%s%d%d", op, &a, &b);
        pa = find(a), pb = find(b);
        if (op[0] == 'M') {
            if (pa != pb) p[pa] = pb;
        } else {
            if (pa == pb) printf("Yes\n");
            else printf("No\n");
        }
    }
    return 0;
}

int find(int x) {
    if (p[x] != x) p[x] = find(p[x]);
    return p[x];
}

File: Data_structure/836/disjoint_set.py
def find(x: int, p: list) -> int:
    root = x
    while p[root] != root:
        root = p[root]
    while p[x] != x:
        t = p[x]
        p[x] = root
        x = p[t]
    return root
    
if __name__ == "__main__":
    n, m = map(int, input().split())
    p = [i for i in range(n + 1)]
    for _ in range(m):
        line = input().split()
        op, x, y = line[0], int(line[1]), int(line[2])
        px, py = find(x, p), find(y, p)
        if op == "M":
            if px != py:
                p[px] = py
        else:
            if px == py:
                print("Yes")
            else:
                print("No")
